### PR TITLE
DecoderAttention with cross attention and cache

### PR DESCRIPTION
**Pull request by wangyems**
**Description**: Describe your changes.
- support decoder attention in fluency bart model
- accept (s, b, h) as input shape
- support cross attention: takes in query and key with weights split into q_weight and kv_weight
- support keys_padding_mask
- support using key/value cache and output cache

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.
For Fluency model

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/bert/decoder_attention.cc [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+  const Tensor* kv_weights(context->Input<Tensor>(3));
+  const Tensor* bias(context->Input<Tensor>(4));
+  const Tensor* key_padding_mask(context->Input<Tensor>(5));
+  const Tensor* key_cache(context->Input<Tensor>(6));
+  const Tensor* value_cache(context->Input<Tensor>(7));
+  const Tensor* static_kv(context->Input<Tensor>(8));
+  const Tensor* use_past(context->Input<Tensor>(9));
+  const Tensor* has_layer_state(context->Input<Tensor>(10));
+  const Tensor* has_key_padding_mask(context->Input<Tensor>(11));
+  //TODO: check inputs
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/decoder_attention.cc:53]: Will add the input check

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 2]
```diff
@@ -526,6 +526,22 @@ void AttentionTypeAndShapeInference(ONNX_NAMESPACE::InferenceContext& ctx, int p
   }
 }
 
+void DecoderAttentionTypeAndShapeInference(ONNX_NAMESPACE::InferenceContext& ctx) {
```
Comment (wangyems) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:529]: TODO

**Comment by tianleiwu** [comment]
Could you check that it could work with bart ONNX exported from huggingface?
https://github.com/huggingface/transformers/tree/master/examples/onnx/pytorch/translation/bart_onnx

**Comment by wangyems** [comment]
> Could you check that it could work with bart ONNX exported from huggingface? https://github.com/huggingface/transformers/tree/master/examples/onnx/pytorch/translation/bart_onnx

Will add the parity check for the latest huggingface bart.

**Comment by wangyems** [comment]
closing to archive this branch. Will instead send out another PR

**Pull request closed without merging**